src/ess/polarization/base.py

```diff
@@ -58,116 +64,227 @@ class He3TransmissionEmptyGlass(sl.Scope[Cell, sc.DataArray], sc.DataArray):
+    return sc.DataArray(
+        data, coords={'time': time_bands, 'wavelength': wavelength_bands}
+    )
 
 
-def spin_channel(
-    polarizer_spin: CellSpin[Polarizer], analyzer_spin: CellSpin[Analyzer]
-) -> SpinChannel:
+def run_reduction_workflow(
+    run_section: RunSectionLog,
+    wavelength_bands: WavelengthBins,
+) -> ReducedDataByRunSectionAndWavelength:
     """
-    Returns a time series of the combined spin channel (++, +-, -+, --).
+    Run the reduction workflow.
+
+    Note that is it currently not clear if we will wrap the workflow in a function,
+    or assemble a common workflow. The structural details may thus be subject to
+    change.
 
-    This will be used to split the raw detector data into the four spin channels.
+    The reduction workflow must return normalized event data, binned into time and
+    wavelength bins. The time bands define intervals of different meaning, such as
+    sample runs, direct beam runs, and spin states.
     """
-    # TODO In practice, are switching times instant? Do we need to drop events that
-    # occur during switching? How is this marked in the meta data?
-    return SpinChannel()
+    # TODO
+    # Subdivide sample section into smaller intervals, or returnnumerator/denominator
+    # separately? The latter would complicate things when supporting different
+    # kinds of workflows, performing different kinds of normalizations.
+    # We need to be careful when subdividing and (1) exactly preserve existing bounds
+    # and (2) introduced new bounds using some heuristics that yield approximately
+    # equal time intervals (for the sample runs).
+    data = dummy_reduction(
+        time_bands=run_section.coords['time'],
+        wavelength_bands=wavelength_bands,
+    )
+    for name, log in run_section.items():
+        data.coords[name] = log.data
+    return ReducedDataByRunSectionAndWavelength(data)
+
+
+def compute_direct_beam(
+    data: sc.DataArray,
+    q_range: sc.Variable,
+    background_q_range: sc.Variable,
+) -> sc.DataArray:
+    """Compute background-subtracted direct beam function."""
+    start_db = q_range[0]
+    stop_db = q_range[-1]
+    start_bg = background_q_range[0]
+    stop_bg = background_q_range[-1]
+    # The input is binned in time and wavelength, we simply histogram without changes.
+    direct_beam = data.bins['Q', start_db:stop_db].hist()
+    background = data.bins['Q', start_bg:stop_bg].hist()
+    return direct_beam - background
+
+
+ReducedDirectBeamDataNoCell = NewType('ReducedDirectBeamDataNoCell', sc.DataArray)
+
+
+class ReducedDirectBeamData(
+    sl.ScopeTwoParams[Cell, PolarizationState, sc.DataArray], sc.DataArray
+):
+    """Direct beam data for a given cell, as a function of wavelength and time."""
+
+
+def extract_direct_beam(
+    data: ReducedDataByRunSectionAndWavelength,
+) -> ReducedDirectBeamDataNoCell:
+    """Extract direct beam without any cells from direct beam data."""
+    is_direct_beam = ~(
```

Comment:
I think it might carry trough to the computation of the direct-beam function later (above), which would probably return all zeros? @astellhorn what should we do in such cases?

Reply:
The direct beam measurement without cell is needed to calculate the initial nuclear polarization of the cell, if its not measured then this parameter is unknown and the polarization correction cannot be performed with the same workflow. However, the workflow then could be adapted to fit this unknown value. It would be less precise but possible. But still it would mean a slightly different workflow. Hence I propose to write this workflow first assuming all measurements will be acquired, and then afterwards adapt for special cases.